client/components/data/query-products-list/index.ts

```diff
@@ -0,0 +1,57 @@
+import { useEffect } from 'react';
+import { useDispatch } from 'react-redux';
+import { requestProductsList } from 'calypso/state/products-list/actions';
+import { isProductsListFetching, getProductsListType } from 'calypso/state/products-list/selectors';
+import type { CalypsoDispatch } from 'calypso/state/types';
+import type { AppState } from 'calypso/types';
+
+type QueryProductsListProps = {
+	type?: string;
+	currency?: string;
+	persist?: boolean;
+	productSlugList?: string[];
+};
+
+type RequestProps = Omit< QueryProductsListProps, 'productSlugList' > & {
+	product_slugs?: string[];
+};
+
+const request =
+	( { persist, ...props }: RequestProps ) =>
+	( dispatch: CalypsoDispatch, getState: () => AppState ) => {
+		if (
+			isProductsListFetching( getState() ) ||
+			( persist && props.type === getProductsListType( getState() ) )
+		) {
+			return;
+		}
+
+		dispatch( requestProductsList( props as any ) );
+	};
+
+export function useQueryProductsList( {
+	type = 'all',
+	currency,
+	persist,
+	productSlugList,
+}: QueryProductsListProps = {} ) {
+	const dispatch = useDispatch();
+
+	// Only runs on mount.
+	useEffect( () => {
+		// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+		// @ts-ignore
+		dispatch( request( { type, currency, persist, product_slugs: productSlugList } ) );
```

Comment:
This TS error can be fixed by typing the `dispatch` function as `CalypsoDispatch`, then it will be able to dispatch the thunk that `request` returns. Create `dispatch` like this:
```ts
const dispatch = useDispatch< CalypsoDispatch >();
```

Reply:
Yup, you're right it's a quick fix.